File: java-collections-maps/src/main/java/com/baeldung/map/hashCode/Money.java
package com.baeldung.map.hashcode;

class Money {

    int amount;
    String currencyCode;

    Money(int amount, String currencyCode) {
        this.amount = amount;
        this.currencyCode = currencyCode;
    }

    @Override
    public boolean equals(Object o) {
        if (o == this)
            return true;
        if (!(o instanceof Money))
            return false;
        Money other = (Money)o;
        return this.amount == other.amount
          && this.currencyCode == other.currencyCode;
    }

    @Override
    public int hashCode() {
        int result = 17;
        result = 31 * result + amount;
        result = 31 * result + currencyCode.hashCode();
        return result;
    }

}

File: core-kotlin-modules/core-kotlin-lang/src/main/kotlin/com/baeldung/operators/Money.kt
package com.baeldung.operators

import java.math.BigDecimal

enum class Currency {
    DOLLARS, EURO
}

class Money(val amount: BigDecimal, val currency: Currency) : Comparable<Money> {

    override fun compareTo(other: Money): Int =
            convert(Currency.DOLLARS).compareTo(other.convert(Currency.DOLLARS))

    fun convert(currency: Currency): BigDecimal = TODO()

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Money) return false

        if (amount != other.amount) return false
        if (currency != other.currency) return false

        return true
    }

    override fun hashCode(): Int {
        var result = amount.hashCode()
        result = 31 * result + currency.hashCode()
        return result
    }
}